File: 백준/Silver/2792. 보석 상자/보석 상자.cc
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;
ll n, m, a[300004], ret = 1e18;
bool check(ll mid){
    ll num = 0;
    for(int i = 0; i < m; i++){
        num += a[i] / mid;
        if(a[i] % mid) num++;
    }
    return n >= num;
}
int main(){
    ios::sync_with_stdio(0);cin.tie(0);cout.tie(0);
    cin >> n >> m;
    ll lo = 1, hi = 1, mid;
    for(int i = 0; i < m; i++) cin >> a[i], hi = max(hi, a[i]);
    while(lo <= hi){
        mid = (lo + hi) / 2;
        if(check(mid)){
            ret = min(ret, mid);
            hi = mid - 1;
        }else lo = mid + 1;
    }
    cout << ret << '\n';
}

File: 백준/Silver/2792. 보석 상자/보석 상자.py
import sys
input = sys.stdin.readline
n,m  = map(int, input().split())
nums = []
for _ in range(m):
    nums.append(int(input()))
l, r = 1, max(nums)
ret = int(1e9)
def check(mid):
    global n
    cnt = 0
    for num in nums:
        temp = num // mid if num % mid == 0 else num // mid + 1 
        cnt += temp
        if cnt > n:
            return False
    return True
while l <= r:
    mid = (l + r) // 2
    if check(mid):
        r = mid - 1
        ret = min(ret, mid)
    else:
        l = mid + 1
print(ret)